aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp

```diff
@@ -0,0 +1,556 @@
+   */
+  bool is_safe_to_get_storage_as_tensor(const NestedTensorImpl* tensor) {
+    const int64_t* tensor_offsets_ptr =
+        tensor->get_storage_offsets().data_ptr<int64_t>();
+    const Tensor& tensor_sizes = tensor->get_nested_sizes();
+    const Tensor& tensor_strides = tensor->get_nested_strides();
+
+    const int64_t n_tensors = tensor_strides.size(0);
+    const int64_t n_dims = tensor_strides.size(1);
+
+    if (n_tensors <= 1) {
+      return true;
+    }
+
+    int64_t* previous_tensor_stride = tensor_strides.data_ptr<int64_t>();
+    // Check initially that they are in strictly descending order
+    for (int i{1}; i < n_dims; i++) {
+      if (previous_tensor_stride[i - 1] <= previous_tensor_stride[i]) {
+        return false;
+      }
+    }
+    // Check that each tensor i in the nested tensor has the same strides
+    auto tensor_stride_0 = tensor_strides.stride(0);
+
+    for (int i{1}; i < n_tensors; i++) {
+      for (const int64_t j : c10::irange(n_dims)) {
+        if (previous_tensor_stride[j] !=
+            previous_tensor_stride[i * tensor_stride_0 + j]) {
+          return false;
+        }
+      }
+    }
+    // Check the offsets are a constant multiple from the previous numels
+    const int64_t* tensor_size_ptr = tensor_sizes.data_ptr<int64_t>();
+    const int64_t* tensor_stride_ptr = tensor_strides.data_ptr<int64_t>();
+
+    int64_t numel_0 = (tensor_size_ptr[0] * tensor_stride_ptr[0]);
+    TORCH_INTERNAL_ASSERT(numel_0 > 0, "numels must be positive!");
+
+    int64_t offset_constant =
+        (tensor_offsets_ptr[1] - tensor_offsets_ptr[0]) / numel_0;
+    for (int64_t i = 2; i < n_tensors; i++) {
+      // TODO: When 0 seq_len nested tensors are allowed we need to guard
+      // against this
+      int64_t previous_numel = tensor_size_ptr[(i - 1) * tensor_stride_0] *
+          tensor_stride_ptr[(i - 1) * tensor_stride_0];
+      TORCH_INTERNAL_ASSERT(previous_numel > 0, "numels must be positive!");
+      int64_t current_offset_constant =
+          (tensor_offsets_ptr[i] - tensor_offsets_ptr[i - 1]) / previous_numel;
+      if (current_offset_constant != offset_constant) {
+        return false;
+      }
+    }
+    // Congrats you made it!
+    return true;
+  }
+
+  /**
+   * Process an individual NestedTensor to reshape and view as a DenseTensor
+   * Generally the approach for q, k, v is to
+   * (1) get the storage of the contiguous nested tensor
+   * (2) view as shape {output_batch_size, {*}_t.size(1), output_num_heads,
+   * head_dim_{*}}, and stride {0, nnz_{*}_stride, head_{*}_stride,
+   * head_dim_stride} where head_{*}_stride is 0 if
+   * {*}_num_heads_needs_broadcast (3) collapse the first two dims by reshaping
+   * to {Nnz_{*}, output_num_heads, head_dim_{*}} if {*}_t.size(1) (i.e. the
+   * seq_len is 1), the reshape should be a view and should not incur a copy
+   *  dense tensor without getting the storage
+   */
+  at::Tensor view_as_dense(
```

Comment:
I abstracted this out since we need it for both the forward preprocessing(broadcasted or not) and backwards preprocessing, I also think it makes it much cleaner